fix(extui): only append messages exceeding 'cmdheight' to "more"

luukvbaal opened the pull request:
**fix(extui): only append messages exceeding 'cmdheight' to "more"**

Problem:  8defe1af declared the "more" window the most convenient place to
          route messages to if it is already open for msg.pos == 'cmd'.
          In usage, this doesn't appear to be the case. Appending messages
          as added in that commit is still useful, but should only be done
          for messages that spill 'cmdheight'.
Solution: Only append messages exceeding 'cmdheight' to the more window.
          To do this, instead of immediately writing to the more buffer,
          write to the cmd buffer and calculate its height. Then copy the
          text and its highlights to the more buffer.

Fix #34372
Fix #34374

**docs(extui): rename box->msg, more->pager, prompt->dialog**

Includes breaking changes to the `opts` layout. "box" doesn't really
describe anything other than a floating window so was an unwanted synonym.

luukvbaal committed:
fix(extui): only append messages exceeding 'cmdheight' to "more"

Problem:  8defe1a declared the "more" window the most convenient place to
          route messages to if it is already open for msg.pos == 'cmd'.
          In usage, this doesn't appear to be the case. Appending messages
          as added in that commit is still useful, but should only be done
          for messages that spill 'cmdheight'.
Solution: Only append messages exceeding 'cmdheight' to the more window.
          To do this, instead of immediately writing to the more buffer,
          write to the cmd buffer and calculate its height. Then copy the
          text and its highlights to the more buffer.

phanen commented:
type `:ls`, `:q`, `:com`:
```
vim.schedule callback: /home/phan/b/neovim/runtime/lua/vim/_extui/messages.lua:193: Invalid 'end_col': out of range
stack traceback:
	[C]: in function 'nvim_buf_set_extmark'
	/home/phan/b/neovim/runtime/lua/vim/_extui/messages.lua:193: in function </home/phan/b/neovim/runtime/lua/vim/_extui/messages.lua:183>
```

phanen commented:
Two behaviors different from builtin one:

`:echo 1 2`
<details>

builtin:
```
1 2
```
extui:
```
1
2
```
</details>

`:echon 1 2`, `:q`, `g<`

<details>

builtin:
```
12
```

extui:
```
1
2
```
</details>

phanen commented:
`nvim --clean --cmd "lua require('vim._extui').enable({})" +"echo globpath(&rtp, '*') | com"`
```
vim.schedule callback: /home/phan/b/neovim/runtime/lua/vim/_extui/messages.lua:255: Invalid 'end_col': out of range
stack traceback:
	[C]: in function 'nvim_buf_set_text'
	/home/phan/b/neovim/runtime/lua/vim/_extui/messages.lua:255: in function 'show_msg'
	/home/phan/b/neovim/runtime/lua/vim/_extui/messages.lua:378: in function 'handler'
	/home/phan/b/neovim/runtime/lua/vim/_extui.lua:50: in function 'fn'
	/home/phan/b/neovim/runtime/lua/vim/_editor.lua:367: in function </home/phan/b/neovim/runtime/lua/vim/_editor.lua:366>
```

luukvbaal commented:
> Two behaviors different from builtin one:

Fixed by #34393

> nvim --clean --cmd "lua require('vim._extui').enable({})" +"echo globpath(&rtp, '*') | com"

Thanks I'll look into it.

luukvbaal commented:
This now also addresses #34372, and https://github.com/neovim/neovim/pull/34373#discussion_r2134845380. I went with a dot instead of a hyphen for the buffer name to continue the convention set for augroups/namespaces, or is there a reason a hyphen is better for buffer names @justinmk?

Let's also rename the more window to "pager" here, I think that makes more sense as a buffer name. It's more universally recognized, and it's already a synonym for `:h more-prompt`.

luukvbaal committed:
docs(extui): rename box->msg, more->pager, prompt->dialog

Includes breaking changes to the `opts` layout. "box" doesn't really
describe anything other than a floating window so was an unwanted synonym.

justinmk commented:
> I went with a dot instead of a hyphen for the buffer name to continue the convention set for augroups/namespaces, or is there a reason a hyphen is better for buffer names

For a URI scheme, a dot might be strange since it looks like a TLDN? In general, a URI-like name seems like a good practice. It allows us to clearly namespace names that are owned by Nvim. (`nvim://...`, `nvim-pager://...`)

> Let's also rename the more window to "pager" here, I think that makes more sense as a buffer name.

Yeah. "more" is a historical detail. I don't see "pager" in the current PR diff though.

luukvbaal commented:
> For a URI scheme, a dot might be strange since it looks like a TLDN? In general, a URI-like name seems like a good practice. It allows us to clearly namespace names that are owned by Nvim. (nvim://..., nvim-pager://...)

Yeah I was just thinking of this as a ('buftype' = "nofile") identifier for wherever it shows up. Not familiar with how or where having a clearly defined URI for this would help, but happy to oblige.

justinmk commented:
> prefix `Message` for the message buffer (so `[Message <Box, Pager, Prompt>]`), because `[Prompt]` already exists for the prompt 'buftype'.

- `[Pager]` is too good to pass up, let's stick with that :)
- The "Box" could be named `[Messages]` (and we can use that for the ["Phase 2" overlay](https://github.com/neovim/neovim/issues/34281) too).
- Is the "Prompt" for things like `inputlist()`? Could be named `[Dialog]` or something.

> Not sure whether we can change that to `[Command Window]`

we could try it, but probably not worth the trouble.

> or what else to name the cmdline buffer just so that it doesn't conflict with that (`[Cmd]`, `[Cmdline]`).

`[Cmd]` sounds good

luukvbaal commented:
Good suggestions, I think. Also using those for the 'filetype' then.

phanen commented:
`:echo 1 2 | echon 1 2` then `g<` behave different(`1 2\n12`) from builtin(`1 212`)

Maybe `msg_history_show` need a `append` for each entry:

<details>

```diff
diff --git a/runtime/lua/vim/_extui/messages.lua b/runtime/lua/vim/_extui/messages.lua
index 107044edc7..e25ebb5723 100644
--- a/runtime/lua/vim/_extui/messages.lua
+++ b/runtime/lua/vim/_extui/messages.lua
@@ -414,7 +414,7 @@ function M.msg_ruler(content)
   set_virttext('last')
 end
 
----@alias MsgHistory [string, MsgContent]
+---@alias MsgHistory [string, MsgContent, boolean]
 --- Open the message history in the pager.
 ---
 ---@param entries MsgHistory[]
@@ -425,7 +425,7 @@ function M.msg_history_show(entries)
 
   api.nvim_buf_set_lines(ext.bufs.pager, 0, -1, false, {})
   for i, entry in ipairs(entries) do
-    M.show_msg('pager', entry[2], i == 1, false)
+    M.show_msg('pager', entry[2], i == 1, entry[3])
   end
 
   M.set_pos('pager')
diff --git a/src/nvim/message.c b/src/nvim/message.c
index c776c39c64..71d2c4a6a0 100644
--- a/src/nvim/message.c
+++ b/src/nvim/message.c
@@ -1053,6 +1053,7 @@ static void msg_hist_add_multihl(HlMessage msg, bool temp)
   entry->kind = msg_ext_kind;
   entry->prev = msg_hist_last;
   entry->next = NULL;
+  entry->append = msg_ext_append;
 
   if (msg_hist_first == NULL) {
     msg_hist_first = entry;
@@ -1206,6 +1207,7 @@ void ex_messages(exarg_T *eap)
         ADD(content, ARRAY_OBJ(content_entry));
       }
       ADD(entry, ARRAY_OBJ(content));
+      ADD(entry, BOOLEAN_OBJ(p->append));
       ADD(entries, ARRAY_OBJ(entry));
     }
     if (redirecting() || !ui_has(kUIMessages)) {
diff --git a/src/nvim/message_defs.h b/src/nvim/message_defs.h
index a65f0cd97a..f4ca7f5f2e 100644
--- a/src/nvim/message_defs.h
+++ b/src/nvim/message_defs.h
@@ -18,4 +18,5 @@ typedef struct msg_hist {
   HlMessage msg;          ///< Highlighted message.
   const char *kind;       ///< Message kind (for msg_ext)
   bool temp;              ///< Temporary message since last command ("g<")
+  bool append;            ///< msg_ext_append
 } MessageHistoryEntry;
```

<details>

luukvbaal commented:
Or we could just append the messages internally, that would ensure the history event is formatted properly too. I opted to add the `append` parameter to just be able to set it accordingly in case other message sources need to append but I think `:echon` might be the only one. Not too late to revert abb40ece at this point...

luukvbaal commented on runtime/lua/vim/_extui/shared.lua:
```diff
@@ -33,7 +33,7 @@ local tab = 0
 function M.tab_check_wins()
   local curtab = api.nvim_get_current_tabpage()
-  for _, type in ipairs({ 'box', 'cmd', 'more', 'prompt' }) do
+  for _, type in ipairs({ 'box', 'cmd', 'pager', 'dialog' }) do
```
Could work but in that case we should probably also rename the `cfg.msg.pos == 'box'` config we already have for it, which then becomes a bit nondescript. Still OK as long as its document it I guess.

phanen commented on runtime/lua/vim/_extui/messages.lua:
```diff
@@ -296,8 +296,9 @@ function M.show_msg(tar, content, replace_last, append, more)
       local h = api.nvim_win_text_height(ext.wins.cmd, {})
-      if (more or not api.nvim_win_get_config(ext.wins.cmd).hide) and h.all > ext.cmdheight then
-        msg_to_more(tar)
+      local want_pager = pager or will_pager or not api.nvim_win_get_config(ext.wins.pager).hide
```
No.. the same error can be triggered with:
```
nvim --clean --cmd "lua require('vim._extui').enable({})" +com +ls
```

luukvbaal replied:
Thanks, fixed the source this time.
